### ISSUE TITLE
How to use Mamba for forecasting

### ISSUE THREAD

**Issue by IsmailKhazi** [wzhwzhwzh0921/S-D-Mamba#14]
Hello can you provide a Python Notebook for "How to do forecasting using Mamba?"

How to use Mamba to forecast and compare with LSTM, RNN also.

**Comment by wzhwzhwzh0921** [comment]
You can read Mamba's paper to learn the principles of Mamba Block. In fact, the input and output of Mamba Block are the same as those of classic RNN models such as LSTM and GRU. The input is the original tensor of the BLD shape, and the output is the hidden tensor corresponding to each position of the BLD shape.

**Comment by wzhwzhwzh0921** [comment]
You can read Mamba's paper to learn the principles of Mamba Block. In fact, the input and output of Mamba Block are the same as those of classic RNN models such as LSTM and GRU. The input is the original tensor of the BLD shape, and the output is the hidden tensor (BLD Shape) corresponding to each position.

**Comment by IsmailKhazi** [comment]
But my code is related to bus arrival time which has both categorical and
numerical features. Can you guide me how to do encoding for these. If
needed I can share sample data for reference.

**Comment by TomKoester** [comment]
An little example for forecasting in a notebook would be great @wzhwzhwzh0921

**Comment by IsmailKhazi** [comment]
I just want how to do encoding and train the model. As it is numerical data
in a csv format. Can you guide me with that.

**Comment by TomKoester** [comment]
The data preproccesing for excatly this model i cant help. I just build an Prototype of the iTransformer model last week for TSA. 
For encode categorial data just encode it with numbers. If i get you problem right that would be something like bus:1 train:2 .... and after encoding you can normalize the data with sin/cos.

**Comment by IsmailKhazi** [comment]
Here is my data description

Name  Description  Data Type  Example
service_date  Date on which the trip took place. Service dates run from
around 2:00AM - 1:59:59AM  Date  2019-12-31
route_id  Route Id  String  01
direction_id  Identifies whether the trip is traveling inbound or outbound
String  Inbound
half_trip_id  Identification for the one way trip  Integer  40836717
stop_id  GTFS-compatible stop  Integer  75
time_point_id  The code for the timepoint  String  mit
time_point_order  The order of this timepoint in the trip  Integer  4
point_type  Identifies whether the stop is the starting point, midpoint, or
endpoint for the trip  String  Midpoint
standard_type  Identifies whether the trip should be evaluated on the
scheudle standard or headway standard.  String  Headway
scheduled  The time the trip was scheduled to depart the stop. The
scheduled time should not be used to evaluated reliability if this trip is
evaluated on the headway standard. Only the headway and run time should be
used per the Service Delivery Policy.  Time  12:30 AM
actual  The time the trip actually departed the timepoint.  Time  12:29 AM
scheduled_headway  The scheduled time between the trip and the previous
trip at the stop, in seconds. NULL if this trip is evaluated on the
schedule standard.  Integer  1200
headway  The actual time between the trip and the previous trip at the
stop, in seconds. NULL if this trip is evaluated on the schedule standard
or if this is the last stop on the trip (endpoint). Endpoints are evaluated
by whether the trip runtime is within 120% of the scheduled run time.
Integer  1163

**Comment by ericleonardo** [comment]
You can try **Mambular** (Mamba for Tabular data) new Python package ready to use:

https://github.com/basf/mamba-tabular

**Comment by IsmailKhazi** [comment]
Is it using Mamba concept under the hood

**Comment by ericleonardo** [comment]
Yes, it is Mamba based. Take a look at the description and docs:
https://mambular.readthedocs.io/

> Mambular is a Python package that brings the power of Mamba architectures to tabular data, offering a suite of deep learning models for regression, classification, and distributional regression tasks. Designed with ease of use in mind, Mambular models adhere to scikit-learn's BaseEstimator interface, making them highly compatible with the familiar scikit-learn ecosystem. This means you can fit, predict, and evaluate using Mambular models just as you would with any traditional scikit-learn model, but with the added performance and flexibility of deep learning.

**Comment by IsmailKhazi** [comment]
In the github it is showing an example of Classification. But my use case
is for Regression.

If possible can you make a simple example code snippet for Regression and
Classification algorithm using Mamba

**Comment by ericleonardo** [comment]
Just read the docs. Regression:

https://mambular.readthedocs.io/en/latest/examples/regression.html

**Comment by IsmailKhazi** [comment]
Yeah I got that, My problem is more leaned to Time Series Forecasting. Can
you give some suggestions on how to use it for time series.

**Comment by ericleonardo** [comment]
You can use the regression. In the train_test_split use shuffle=False to avoid leaking future samples in the training.
Please, show how your data is structured. The description you sent above is confused because mixed columns.
You can use print screen and paste here.

> Name  Description  Data Type  Example
> service_date  Date on which the trip took place. Service dates run from
> around 2:00AM - 1:59:59AM  Date  2019-12-31
> route_id  Route Id  String  01

**Comment by TomKoester** [comment]
Is it also appliable for probalistic output to do multilabel classification? @ericleonardo

**Comment by ericleonardo** [comment]
> Is it also appliable for probalistic output to do multilabel classification? @ericleonardo

You can try the Classification with multiclass output. In the example, it is used 4 classes:
https://mambular.readthedocs.io/en/latest/examples/classification.html

> Generate target variable:
> y = np.dot(X, coefficients) + np.random.randn(n_samples)
> Convert y to multiclass by categorizing into quartiles
> y = pd.qcut(y, 4, labels=False)

**Comment by TomKoester** [comment]
@ericleonardo Hey, thanks for the fast reply.
I think you got me wrong there, i need an probalistic output for each class for a mutlilabel prediction. That means each input can be more than just one class.
Example
Input Class1 Class2 Class3
19.6.24    1    0    1

**Comment by IsmailKhazi** [comment]
Here is the link for the data description

https://mbta-massdot.opendata.arcgis.com/datasets/ef464a75666349f481353f16514c06d0/about

**Comment by ericleonardo** [comment]
> @ericleonardo Hey, thanks for the fast reply. I think you got me wrong there, i need an probalistic output for each class for a mutlilabel prediction. That means each input can be more than just one class. Example Input Class1 Class2 Class3 19.6.24 1 0 1

I don't know if it will output class probabilities. You can ask the author:
https://github.com/basf/mamba-tabular
